File: src/liffier/fuzzer.py
"""Core fuzzer: dispatches requests and collects results."""

from __future__ import annotations

import json
import time
from concurrent.futures import ThreadPoolExecutor, as_completed
from dataclasses import asdict, dataclass, field
from pathlib import Path
from typing import Any
from urllib.parse import urljoin

import requests

from liffier.detector import Detection, analyze_response
from liffier.encodings import build_payloads


@dataclass
class FuzzResult:
    """Single fuzzing attempt result."""

    url: str
    payload: str
    encoding: str
    depth: int
    suffix: str
    status_code: int
    content_length: int
    elapsed_ms: int
    detection: Detection
    response_snippet: str = ""
    error: str | None = None

    def to_dict(self) -> dict[str, Any]:
        d = {
            "url": self.url,
            "payload": self.payload,
            "encoding": self.encoding,
            "depth": self.depth,
            "suffix": self.suffix,
            "status_code": self.status_code,
            "content_length": self.content_length,
            "elapsed_ms": self.elapsed_ms,
            "hit": self.detection.hit,
            "confidence": self.detection.confidence,
            "reason": self.detection.reason,
            "matched_signature": self.detection.matched_signature,
            "response_snippet": self.response_snippet,
            "error": self.error,
        }
        return d


@dataclass
class FuzzConfig:
    """Fuzzer configuration."""

    url: str
    target_file: str
    max_depth: int = 10
    encodings: list[str] | None = None
    bypass_suffixes: bool = False
    workers: int = 10
    timeout: int = 10
    delay: float = 0.0
    proxy: str | None = None
    cookies: dict[str, str] = field(default_factory=dict)
    headers: dict[str, str] = field(default_factory=dict)
    method: str = "GET"
    follow_redirects: bool = False
    snippet_length: int = 200


class PathTraversalFuzzer:
    """Fuzz a URL parameter for path traversal vulnerabilities."""

    def __init__(self, config: FuzzConfig):
        self.config = config
        self.session = requests.Session()
        self.session.verify = False
        if config.proxy:
            self.session.proxies = {
                "http": config.proxy,
                "https": config.proxy,
            }
        if config.cookies:
            self.session.cookies.update(config.cookies)
        if config.headers:
            self.session.headers.update(config.headers)
        self.baseline_length: int | None = None

    def _get_baseline(self) -> None:
        """Fetch baseline response for comparison."""
        try:
            test_url = self.config.url + "nonexistent_file_baseline_test_12345"
            resp = self.session.request(
                self.config.method,
                test_url,
                timeout=self.config.timeout,
                allow_redirects=self.config.follow_redirects,
            )
            self.baseline_length = len(resp.text)
        except Exception:
            self.baseline_length = None

    def _fuzz_single(self, payload_info: dict[str, str]) -> FuzzResult:
        """Send a single traversal request."""
        payload = payload_info["payload"]
        url = self.config.url + payload

        try:
            resp = self.session.request(
                self.config.method,
                url,
                timeout=self.config.timeout,
                allow_redirects=self.config.follow_redirects,
            )
            detection = analyze_response(
                status_code=resp.status_code,
                body=resp.text,
                target_file=self.config.target_file,
                baseline_length=self.baseline_length,
            )
            snippet = resp.text[:self.config.snippet_length].strip()
            return FuzzResult(
                url=url,
                payload=payload,
                encoding=payload_info["encoding"],
                depth=int(payload_info["depth"]),
                suffix=payload_info["suffix"],
                status_code=resp.status_code,
                content_length=len(resp.content),
                elapsed_ms=int(resp.elapsed.total_seconds() * 1000),
                detection=detection,
                response_snippet=snippet,
            )
        except requests.exceptions.Timeout:
            return FuzzResult(
                url=url,
                payload=payload,
                encoding=payload_info["encoding"],
                depth=int(payload_info["depth"]),
                suffix=payload_info["suffix"],
                status_code=0,
                content_length=0,
                elapsed_ms=self.config.timeout * 1000,
                detection=Detection(hit=False, confidence="high", reason="Timeout"),
                error="Request timed out",
            )
        except Exception as exc:
            return FuzzResult(
                url=url,
                payload=payload,
                encoding=payload_info["encoding"],
                depth=int(payload_info["depth"]),
                suffix=payload_info["suffix"],
                status_code=0,
                content_length=0,
                elapsed_ms=0,
                detection=Detection(hit=False, confidence="high", reason=str(exc)),
                error=str(exc),
            )

    def run(self, callback: Any = None) -> list[FuzzResult]:
        """Execute the fuzzing campaign.

        Args:
            callback: Optional callable(FuzzResult) invoked after each request.
        """
        payloads = build_payloads(
            target_file=self.config.target_file,
            max_depth=self.config.max_depth,
            encoding_names=self.config.encodings,
            use_bypass_suffixes=self.config.bypass_suffixes,
        )

        self._get_baseline()

        results: list[FuzzResult] = []

        with ThreadPoolExecutor(max_workers=self.config.workers) as pool:
            futures = {}
            for i, payload_info in enumerate(payloads):
                if self.config.delay and i > 0:
                    time.sleep(self.config.delay)
                fut = pool.submit(self._fuzz_single, payload_info)
                futures[fut] = payload_info

            for fut in as_completed(futures):
                result = fut.result()
                results.append(result)
                if callback:
                    callback(result)

        return results


def export_results(results: list[FuzzResult], path: Path, fmt: str = "json") -> None:
    """Write results to a file."""
    records = [r.to_dict() for r in results]

    if fmt == "json":
        path.write_text(json.dumps(records, indent=2), encoding="utf-8")
    elif fmt == "jsonl":
        with path.open("w", encoding="utf-8") as fh:
            for rec in records:
                fh.write(json.dumps(rec) + "\n")
    elif fmt == "csv":
        import csv

        if not records:
            path.write_text("", encoding="utf-8")
            return
        with path.open("w", newline="", encoding="utf-8") as fh:
            writer = csv.DictWriter(fh, fieldnames=records[0].keys())
            writer.writeheader()
            writer.writerows(records)
    else:
        raise ValueError(f"Unsupported format: {fmt}")

File: internal/fuzzer/fuzzer.go
package fuzzer

import (
	"crypto/tls"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"sync"
	"time"

	"github.com/momenbasel/liffier/internal/detector"
	"github.com/momenbasel/liffier/internal/encoding"
)

// Result is the outcome of a single fuzz attempt.
type Result struct {
	URL             string
	Payload         string
	Encoding        string
	Depth           int
	Suffix          string
	StatusCode      int
	ContentLength   int
	ElapsedMs       int64
	Detection       detector.Detection
	ResponseSnippet string
	Error           string
}

// Config holds fuzzer settings.
type Config struct {
	URL              string
	TargetFile       string
	MaxDepth         int
	Encodings        []string
	BypassSuffixes   bool
	Workers          int
	Timeout          time.Duration
	Delay            time.Duration
	Proxy            string
	Cookies          string
	Headers          map[string]string
	Method           string
	FollowRedirects  bool
	SnippetLength    int
}

// Fuzzer runs path traversal payloads against a target.
type Fuzzer struct {
	config         Config
	client         *http.Client
	baselineLength int
}

// New creates a Fuzzer from the given config.
func New(cfg Config) *Fuzzer {
	if cfg.Method == "" {
		cfg.Method = "GET"
	}
	if cfg.SnippetLength == 0 {
		cfg.SnippetLength = 200
	}
	if cfg.Workers == 0 {
		cfg.Workers = 10
	}
	if cfg.Timeout == 0 {
		cfg.Timeout = 10 * time.Second
	}

	transport := &http.Transport{
		TLSClientConfig:     &tls.Config{InsecureSkipVerify: true},
		MaxIdleConnsPerHost: cfg.Workers,
	}
	if cfg.Proxy != "" {
		proxyURL, err := url.Parse(cfg.Proxy)
		if err == nil {
			transport.Proxy = http.ProxyURL(proxyURL)
		}
	}

	client := &http.Client{
		Transport: transport,
		Timeout:   cfg.Timeout,
	}
	if !cfg.FollowRedirects {
		client.CheckRedirect = func(req *http.Request, via []*http.Request) error {
			return http.ErrUseLastResponse
		}
	}

	return &Fuzzer{config: cfg, client: client}
}

func (f *Fuzzer) fetchBaseline() {
	testURL := f.config.URL + "nonexistent_baseline_12345"
	req, err := f.buildRequest(testURL)
	if err != nil {
		return
	}
	resp, err := f.client.Do(req)
	if err != nil {
		return
	}
	body, _ := io.ReadAll(resp.Body)
	resp.Body.Close()
	f.baselineLength = len(body)
}

func (f *Fuzzer) buildRequest(targetURL string) (*http.Request, error) {
	req, err := http.NewRequest(f.config.Method, targetURL, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("User-Agent", "Mozilla/5.0 (compatible; liffier/2.0)")
	if f.config.Cookies != "" {
		req.Header.Set("Cookie", f.config.Cookies)
	}
	for k, v := range f.config.Headers {
		req.Header.Set(k, v)
	}
	return req, nil
}

func (f *Fuzzer) fuzzOne(p encoding.Payload) Result {
	targetURL := f.config.URL + p.Value

	req, err := f.buildRequest(targetURL)
	if err != nil {
		return Result{
			URL: targetURL, Payload: p.Value, Encoding: p.Encoding,
			Depth: p.Depth, Suffix: p.Suffix, Error: err.Error(),
			Detection: detector.Detection{Confidence: "high", Reason: err.Error()},
		}
	}

	start := time.Now()
	resp, err := f.client.Do(req)
	elapsed := time.Since(start).Milliseconds()

	if err != nil {
		return Result{
			URL: targetURL, Payload: p.Value, Encoding: p.Encoding,
			Depth: p.Depth, Suffix: p.Suffix, ElapsedMs: elapsed,
			Error: err.Error(),
			Detection: detector.Detection{Confidence: "high", Reason: err.Error()},
		}
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	bodyStr := string(body)

	det := detector.Analyze(resp.StatusCode, bodyStr, f.config.TargetFile, f.baselineLength)

	snippet := bodyStr
	if len(snippet) > f.config.SnippetLength {
		snippet = snippet[:f.config.SnippetLength]
	}

	return Result{
		URL:             targetURL,
		Payload:         p.Value,
		Encoding:        p.Encoding,
		Depth:           p.Depth,
		Suffix:          p.Suffix,
		StatusCode:      resp.StatusCode,
		ContentLength:   len(body),
		ElapsedMs:       elapsed,
		Detection:       det,
		ResponseSnippet: snippet,
	}
}

// Run executes the fuzzing campaign and calls callback for each result.
func (f *Fuzzer) Run(callback func(Result)) []Result {
	payloads := encoding.BuildPayloads(
		f.config.TargetFile,
		f.config.MaxDepth,
		f.config.Encodings,
		f.config.BypassSuffixes,
	)

	f.fetchBaseline()

	results := make([]Result, 0, len(payloads))
	var mu sync.Mutex

	sem := make(chan struct{}, f.config.Workers)
	var wg sync.WaitGroup

	for i, p := range payloads {
		if f.config.Delay > 0 && i > 0 {
			time.Sleep(f.config.Delay)
		}

		wg.Add(1)
		sem <- struct{}{}

		go func(payload encoding.Payload) {
			defer wg.Done()
			defer func() { <-sem }()

			result := f.fuzzOne(payload)
			mu.Lock()
			results = append(results, result)
			mu.Unlock()

			if callback != nil {
				callback(result)
			}
		}(p)
	}

	wg.Wait()
	return results
}

// PayloadCount returns how many payloads will be generated.
func (f *Fuzzer) PayloadCount() int {
	return len(encoding.BuildPayloads(
		f.config.TargetFile,
		f.config.MaxDepth,
		f.config.Encodings,
		f.config.BypassSuffixes,
	))
}

// FormatStatusCode returns a display string for a status code.
func FormatStatusCode(code int) string {
	if code == 0 {
		return "ERR"
	}
	return fmt.Sprintf("%d", code)
}
